File: .defi-compute.js
const fs = require('fs');

// Chains
const chains = JSON.parse(fs.readFileSync('.chains.json'));
const totalTvl = chains.reduce((s,c)=>s+(c.tvl||0),0);
console.log(`TVL_TOTAL: $${(totalTvl/1e9).toFixed(2)}B`);
const chainsSorted = [...chains].sort((a,b)=>(b.tvl||0)-(a.tvl||0));
console.log("Top 10 chains:");
chainsSorted.slice(0,10).forEach(c=>console.log(`  ${c.name}: $${((c.tvl||0)/1e9).toFixed(2)}B 1d=${c.change_1d}% 7d=${c.change_7d}%`));

const sumW = chains.reduce((s,c)=>s+(c.tvl||0)*(c.change_1d||0)/100,0);
const tvlD = totalTvl ? sumW/totalTvl*100 : 0;
console.log(`TVL_1D_PCT: ${tvlD.toFixed(2)}%`);
const sumW7 = chains.reduce((s,c)=>s+(c.tvl||0)*(c.change_7d||0)/100,0);
const tvl7D = totalTvl ? sumW7/totalTvl*100 : 0;
console.log(`TVL_7D_PCT: ${tvl7D.toFixed(2)}%`);

console.log("");
console.log("=== Chain Movers (|1d|>=5%, tvl>=$500M) ===");
const movers = chains.filter(c=>(c.tvl||0)>=500e6 && Math.abs(c.change_1d||0)>=5);
console.log("UP:");
[...movers].sort((a,b)=>(b.change_1d||0)-(a.change_1d||0)).slice(0,5).forEach(c=>console.log(`  ${c.name}: $${((c.tvl||0)/1e9).toFixed(2)}B 1d=${c.change_1d}% 7d=${c.change_7d}%`));
console.log("DOWN:");
[...movers].sort((a,b)=>(a.change_1d||0)-(b.change_1d||0)).slice(0,5).forEach(c=>console.log(`  ${c.name}: $${((c.tvl||0)/1e9).toFixed(2)}B 1d=${c.change_1d}% 7d=${c.change_7d}%`));

// Protocols
console.log("");
const prots = JSON.parse(fs.readFileSync('.protocols.json'));
console.log(`PROTOCOLS_COUNT: ${prots.length}`);
const mp = prots.filter(p=>(p.tvl||0)>=100e6 && Math.abs(p.change_1d||0)>=10);
console.log("Protocol movers (|1d|>=10%, tvl>=$100M, top 12):");
mp.sort((a,b)=>Math.abs(b.change_1d||0)-Math.abs(a.change_1d||0)).slice(0,12).forEach(p=>{
  console.log(`  ${p.name} (${p.category}/${p.chain}): $${((p.tvl||0)/1e9).toFixed(3)}B 1d=${p.change_1d}% 7d=${p.change_7d}%`);
});

// Fees
console.log("");
const fees = JSON.parse(fs.readFileSync('.fees.json'));
console.log(`FEES_TOTAL_24H: $${(fees.total24h/1e6).toFixed(1)}M  total7d=$${(fees.total7d/1e6).toFixed(1)}M  c1d=${fees.change_1d}%  c7d=${fees.change_7d}%`);
const fps = fees.protocols||[];
const fpsSorted = [...fps].sort((a,b)=>(b.total24h||0)-(a.total24h||0));
console.log("Top 10 fees protocols (24h):");
fpsSorted.slice(0,10).forEach(p=>{
  const t24 = p.total24h||0, t7 = p.total7d||0, avg = t7?t7/7:0;
  const delta = avg ? (t24-avg)/avg*100 : 0;
  console.log(`  ${p.name}: 24h=$${(t24/1e6).toFixed(2)}M 7d=$${(t7/1e6).toFixed(1)}M  vs7d_avg=${delta>=0?'+':''}${delta.toFixed(1)}%  c1d=${p.change_1d}%  c7d=${p.change_7d}%`);
});

console.log("");
console.log("=== Fees beating TVL ===");
const protByName = Object.fromEntries(prots.map(p=>[p.name,p]));
const cands = [];
for (const fp of fps) {
  const pp = protByName[fp.name];
  if (!pp) continue;
  const fc7 = fp.change_7d, tc7 = pp.change_7d;
  if (fc7==null || tc7==null) continue;
  if (fc7>20 && tc7<5 && (fp.total24h||0)>=1e6) {
    cands.push({n:fp.name, fc7, tc7, t24:fp.total24h, tvl:pp.tvl});
  }
}
cands.sort((a,b)=>b.fc7-a.fc7).slice(0,10).forEach(c=>{
  console.log(`  ${c.n}: fees_7d=${c.fc7>=0?'+':''}${c.fc7.toFixed(1)}% / tvl_7d=${c.tc7>=0?'+':''}${c.tc7.toFixed(1)}% / 24h_fees=$${(c.t24/1e6).toFixed(2)}M / tvl=$${(c.tvl/1e9).toFixed(2)}B`);
});

// DEX vols
console.log("");
const dexs = JSON.parse(fs.readFileSync('.dexs.json'));
console.log(`DEX_TOTAL_24H: $${(dexs.total24h/1e9).toFixed(2)}B  c1d=${dexs.change_1d}%  c7d=${dexs.change_7d}%`);
const dps = dexs.protocols||[];
[...dps].sort((a,b)=>(b.total24h||0)-(a.total24h||0)).slice(0,6).forEach(p=>{
  console.log(`  ${p.name}: $${((p.total24h||0)/1e9).toFixed(2)}B  c1d=${p.change_1d}%  c7d=${p.change_7d}%`);
});

// Stables
console.log("");
const stables = JSON.parse(fs.readFileSync('.stables.json'));
const peggs = stables.peggedAssets||[];
let totalSupply=0, totalPrev=0, totalPrevW=0;
for (const s of peggs) {
  totalSupply += (s.circulating?.peggedUSD)||0;
  totalPrev += (s.circulatingPrevDay?.peggedUSD)||0;
  totalPrevW += (s.circulatingPrevWeek?.peggedUSD)||0;
}
const stableD = totalPrev? (totalSupply-totalPrev)/totalPrev*100 : 0;
const stableW = totalPrevW? (totalSupply-totalPrevW)/totalPrevW*100 : 0;
console.log(`STABLES_TOTAL: $${(totalSupply/1e9).toFixed(1)}B  1d=${stableD>=0?'+':''}${stableD.toFixed(2)}%  7d=${stableW>=0?'+':''}${stableW.toFixed(2)}%`);

console.log("Top stables by 1d (|1d|>=1%, supply>=$500M):");
const sm = [];
for (const s of peggs) {
  const cc = (s.circulating?.peggedUSD)||0;
  const cp = (s.circulatingPrevDay?.peggedUSD)||0;
  if (cc < 500e6 || !cp) continue;
  const pct = (cc-cp)/cp*100;
  if (Math.abs(pct)>=1) sm.push({sym:s.symbol, cc, pct});
}
sm.sort((a,b)=>Math.abs(b.pct)-Math.abs(a.pct)).slice(0,10).forEach(m=>console.log(`  ${m.sym}: $${(m.cc/1e9).toFixed(2)}B  1d=${m.pct>=0?'+':''}${m.pct.toFixed(2)}%`));

// Yields
console.log("");
const pools = JSON.parse(fs.readFileSync('.pools.json')).data;
console.log(`POOLS_TOTAL: ${pools.length}`);

// Real yield
const real = [];
for (const p of pools) {
  if (p.outlier) continue;
  if ((p.apyBase||0)<=0) continue;
  const apy = p.apy||0, apyR = p.apyReward||0;
  if (apy && (apyR/apy)>=0.5) continue;
  if ((p.predictions?.binnedConfidence||0)<2) continue;
  if ((p.apyMean30d||0)<apy*0.5) continue;
  if ((p.tvlUsd||0)<10e6) continue;
  real.push(p);
}
real.sort((a,b)=>(b.apyBase||0)-(a.apyBase||0));
console.log("Real yield top 12:");
real.slice(0,12).forEach(p=>console.log(`  ${p.symbol} (${p.project}/${p.chain}): apyBase=${p.apyBase.toFixed(2)}% apyReward=${p.apyReward} apy=${p.apy} apyMean30d=${p.apyMean30d} tvl=$${((p.tvlUsd||0)/1e6).toFixed(0)}M bc=${p.predictions?.binnedConfidence}`));

console.log("");
const inc = [];
for (const p of pools) {
  if (p.outlier) continue;
  if ((p.apyReward||0)<=0) continue;
  if ((p.tvlUsd||0)<25e6) continue;
  inc.push(p);
}
inc.sort((a,b)=>(b.apy||0)-(a.apy||0));
console.log("Incentive yield top 10:");
inc.slice(0,10).forEach(p=>console.log(`  ${p.symbol} (${p.project}/${p.chain}): apy=${(p.apy||0).toFixed(1)}% apyReward=${(p.apyReward||0).toFixed(1)}% tvl=$${((p.tvlUsd||0)/1e6).toFixed(0)}M rewards=${(p.rewardTokens||[]).length}`));

File: .defi-compute.py
#!/usr/bin/env python3
import json

# === Chains ===
chains = json.load(open('.chains.json'))
total_tvl = sum(c.get('tvl',0) or 0 for c in chains)
print(f"TVL_TOTAL: ${total_tvl/1e9:.2f}B")
chains_sorted = sorted(chains, key=lambda c: c.get('tvl',0) or 0, reverse=True)
print("Top 10 chains:")
for c in chains_sorted[:10]:
    print(f"  {c.get('name')}: ${(c.get('tvl',0)/1e9):.2f}B 1d={c.get('change_1d')}% 7d={c.get('change_7d')}%")

sum_w = sum((c.get('tvl',0) or 0)*((c.get('change_1d') or 0))/100 for c in chains)
tvl_d_pct = (sum_w/total_tvl)*100 if total_tvl else 0
print(f"TVL_1D_PCT: {tvl_d_pct:.2f}%")
sum_w7 = sum((c.get('tvl',0) or 0)*((c.get('change_7d') or 0))/100 for c in chains)
tvl_7d_pct = (sum_w7/total_tvl)*100 if total_tvl else 0
print(f"TVL_7D_PCT: {tvl_7d_pct:.2f}%")

print()
print("=== Chain Movers (|1d|>=5%, tvl>=$500M) ===")
movers = [c for c in chains if (c.get('tvl',0) or 0) >= 500e6 and abs(c.get('change_1d') or 0) >= 5]
movers_sorted = sorted(movers, key=lambda c: c.get('change_1d',0))
print("DOWN candidates:")
for c in movers_sorted[:5]:
    print(f"  {c.get('name')}: ${(c.get('tvl',0)/1e9):.2f}B 1d={c.get('change_1d')}% 7d={c.get('change_7d')}%")
print("UP candidates:")
for c in sorted(movers, key=lambda c: -c.get('change_1d',0))[:5]:
    print(f"  {c.get('name')}: ${(c.get('tvl',0)/1e9):.2f}B 1d={c.get('change_1d')}% 7d={c.get('change_7d')}%")

# === Protocols ===
print()
prots = json.load(open('.protocols.json'))
print(f"PROTOCOLS_COUNT: {len(prots)}")
mp = [p for p in prots if (p.get('tvl',0) or 0) >= 100e6 and abs(p.get('change_1d') or 0) >= 10]
print("Protocol movers (|1d|>=10%, tvl>=$100M):")
for p in sorted(mp, key=lambda p: -abs(p.get('change_1d',0)))[:10]:
    print(f"  {p.get('name')} ({p.get('category')}/{p.get('chain')}): ${(p.get('tvl',0)/1e9):.3f}B 1d={p.get('change_1d')}% 7d={p.get('change_7d')}%")

# === Fees ===
print()
fees = json.load(open('.fees.json'))
print(f"FEES_TOTAL_24H: ${fees.get('total24h',0)/1e6:.1f}M  total7d=${fees.get('total7d',0)/1e6:.1f}M  change_1d={fees.get('change_1d')}%  change_7d={fees.get('change_7d')}%")
fps = fees.get('protocols',[])
fps_sorted = sorted(fps, key=lambda p: -(p.get('total24h',0) or 0))
print("Top 10 fees protocols (24h):")
for p in fps_sorted[:10]:
    t24 = p.get('total24h',0) or 0
    t7 = p.get('total7d',0) or 0
    avg7d = t7/7 if t7 else 0
    delta = ((t24 - avg7d) / avg7d * 100) if avg7d else 0
    print(f"  {p.get('name')}: 24h=${t24/1e6:.2f}M 7d=${t7/1e6:.1f}M  vs7d_avg={delta:+.1f}%  c1d={p.get('change_1d')}%  c7d={p.get('change_7d')}%")

# Fees beating TVL: fees change_7d > +20% AND TVL change_7d < +5%
print()
print("=== Fees beating TVL (fees 7d>+20% AND TVL 7d<+5%) ===")
# Match fees protocols to /protocols by name
prot_by_name = {p['name']: p for p in prots}
candidates = []
for fp in fps:
    n = fp.get('name')
    pp = prot_by_name.get(n)
    if not pp: continue
    fc7 = fp.get('change_7d')
    tc7 = pp.get('change_7d')
    if fc7 is None or tc7 is None: continue
    if fc7 > 20 and tc7 < 5 and (fp.get('total24h',0) or 0) >= 1e6:
        candidates.append((n, fc7, tc7, fp.get('total24h',0), pp.get('tvl',0)))
for c in sorted(candidates, key=lambda x: -x[1])[:10]:
    print(f"  {c[0]}: fees_7d={c[1]:+.1f}% / tvl_7d={c[2]:+.1f}% / 24h_fees=${c[3]/1e6:.2f}M / tvl=${c[4]/1e9:.2f}B")

# === DEX vols ===
print()
dexs = json.load(open('.dexs.json'))
print(f"DEX_TOTAL_24H: ${dexs.get('total24h',0)/1e9:.2f}B  c1d={dexs.get('change_1d')}%  c7d={dexs.get('change_7d')}%")
dps = dexs.get('protocols',[])
dps_sorted = sorted(dps, key=lambda p: -(p.get('total24h',0) or 0))
print("Top DEXes 24h:")
for p in dps_sorted[:6]:
    print(f"  {p.get('name')}: ${(p.get('total24h',0)/1e9):.2f}B  c1d={p.get('change_1d')}%  c7d={p.get('change_7d')}%")

# === Stables ===
print()
stables = json.load(open('.stables.json'))
peggs = stables.get('peggedAssets',[])
total_supply = 0
for s in peggs:
    cc = s.get('circulating',{}) or {}
    total_supply += cc.get('peggedUSD',0) or 0
total_prev = 0
for s in peggs:
    cp = s.get('circulatingPrevDay',{}) or {}
    total_prev += cp.get('peggedUSD',0) or 0
total_prevw = 0
for s in peggs:
    cpw = s.get('circulatingPrevWeek',{}) or {}
    total_prevw += cpw.get('peggedUSD',0) or 0

stable_d_pct = ((total_supply - total_prev)/total_prev*100) if total_prev else 0
stable_w_pct = ((total_supply - total_prevw)/total_prevw*100) if total_prevw else 0
print(f"STABLES_TOTAL: ${total_supply/1e9:.1f}B  1d={stable_d_pct:+.2f}%  7d={stable_w_pct:+.2f}%")

# Top stables by movement
print("Top stables by 1d change (|1d|>=1%, supply>=$500M):")
movers_s = []
for s in peggs:
    cc = (s.get('circulating',{}) or {}).get('peggedUSD',0) or 0
    cp = (s.get('circulatingPrevDay',{}) or {}).get('peggedUSD',0) or 0
    if cc < 500e6 or not cp: continue
    pct = (cc-cp)/cp*100
    if abs(pct) >= 1:
        movers_s.append((s.get('symbol'), cc, pct))
for m in sorted(movers_s, key=lambda x: -abs(x[2]))[:8]:
    print(f"  {m[0]}: ${m[1]/1e9:.2f}B  1d={m[2]:+.2f}%")

# === Yields ===
print()
pools = json.load(open('.pools.json'))['data']
print(f"POOLS_TOTAL: {len(pools)}")
# Real yield filter
real = []
for p in pools:
    if p.get('outlier'): continue
    if (p.get('apyBase') or 0) <= 0: continue
    apy = p.get('apy') or 0
    apyR = p.get('apyReward') or 0
    apyB = p.get('apyBase') or 0
    if apy and (apyR/apy) >= 0.5: continue
    pred = p.get('predictions',{}) or {}
    bc = pred.get('predictedProbability') and pred.get('binnedConfidence')
    if (pred.get('binnedConfidence') or 0) < 2: continue
    if (p.get('apyMean30d') or 0) < apy * 0.5: continue
    if (p.get('tvlUsd') or 0) < 10e6: continue
    real.append(p)
real_sorted = sorted(real, key=lambda p: -(p.get('apyBase') or 0))
print("Real yield candidates (top 10 by apyBase):")
for p in real_sorted[:10]:
    print(f"  {p.get('symbol')} ({p.get('project')}/{p.get('chain')}): apyBase={p.get('apyBase'):.2f}% apyReward={p.get('apyReward')} tvl=${(p.get('tvlUsd',0)/1e6):.0f}M apyMean30d={p.get('apyMean30d')} bc={p.get('predictions',{}).get('binnedConfidence')}")

# Incentive yield
print()
inc = []
for p in pools:
    if p.get('outlier'): continue
    if (p.get('apyReward') or 0) <= 0: continue
    if (p.get('tvlUsd') or 0) < 25e6: continue
    inc.append(p)
inc_sorted = sorted(inc, key=lambda p: -(p.get('apy') or 0))
print("Incentive yield top 10 by apy:")
for p in inc_sorted[:10]:
    rt = p.get('rewardTokens') or []
    print(f"  {p.get('symbol')} ({p.get('project')}/{p.get('chain')}): apy={p.get('apy'):.1f}% apyReward={p.get('apyReward'):.1f}% tvl=${(p.get('tvlUsd',0)/1e6):.0f}M reward_count={len(rt)}")
